### ISSUE TITLE
GetSyntaxTreeAsync hangs and eats up memory

### ISSUE THREAD

**Issue by aelij** [dotnet/roslyn#12197]
**Version Used**:
2.0.0-beta1-20160607-01
(Works fine on 1.2)

**Steps to Reproduce**:
Run this code:

```csharp
var s = @"
public class SomeClass
{
    public void SomeMethod()
    {
        OtherMethod(

        throw new Exception();
    }
}
";

var w = new AdhocWorkspace();
var p = w.AddProject("X", LanguageNames.CSharp);
var d = p.AddDocument("Y", s);

var syntaxTree = await d.GetSyntaxTreeAsync();
```

**Expected Behavior**:
Syntax tree returned.

**Actual Behavior**:
Value never returns. Process memory keeps growing indefinitely (kept it on for several minutes).

**Comment by gafter** [comment]
@jaredpar Should this be marked urgent? I suspect it can surface as IDE crashing.

**Comment by jaredpar** [comment]
@VSadov can you take a look at this?  I suspect this is tuples related.  

@gafter lets see what the investigation reveals.

**Comment by gafter** [comment]
This may be an error I introduced from an incomplete removal of the throw expression implementation.

**Comment by gafter** [comment]
A community-provided fix for this is in review (#12335).
